### PR TITLE
pytest.warns checks for subclass relationship rather than class equality

### LINKED ISSUE #2151: Should pytest.warns check for issubclass rather than class equality?
`pytest.warns` check for class equality and not subclass relationship. This snippet for example fails:
```py
import warnings

import pytest


with pytest.warns(Warning):
    warnings.warn('user', UserWarning)
```

This was slightly surprising to me since my expectation was that it would be similar to `pytest.raises` where you can check for a base class:
```py
import pytest


with pytest.raises(Exception):
    raise ValueError()
```

What do you guys think? After all you can always use `with pytest.warns(None) as warninfo` and check `issubclass(warninfo[0].message, a_warning_class)` if you want to check a subclass relationship. At the same time, I don't really how it would hurt to check subclass relationship. 

I guess that is just a matter of changing this [line](https://github.com/pytest-dev/pytest/blob/master/_pytest/recwarn.py#L224) by:
```py
if not any(issubclass(r.category, each) for each in self.expected_warning for r in self):
```
I am willing to open a PR about this if you think it is worth it.

### PR DESCRIPTION
**Pull request by lesteve**
This makes it more similar to pytest.raises. Fixes #2151.

### CODE REVIEW

**Comment by nicoddemus** [comment]
Thanks @lesteve for the PR, appreciate it!

I rebased your branch into the latest `features` and pushed to your fork, hope you don't mind.

File: testing/test_recwarn.py [hunk 1]
```diff
@@ -238,6 +238,16 @@ def test_record_only(self):
         assert str(record[0].message) == "user"
         assert str(record[1].message) == "runtime"
 
+    def test_record_by_subclass(self):
```
Comment (nicoddemus) [inline testing/test_recwarn.py]: Could you please also add a test which checks for warning subclasses when the user passes more than one warning to `pytest.warns`? For example:

```python
class Warn1(UserWarning): pass
class Warn2(RuntimeWarning): pass

with pytest.warns((UserWarning, RuntimeWarning)):
   warnings.warn("w1", Warn1)
   warnings.warn("w2", Warn2)

assert len(record) == 2
...
```

**Comment by lesteve** [comment]
> I rebased your branch into the latest features and pushed to your fork, hope you don't mind.

Not at all, I think this is a really nice recent addition in GitHub that maintainers can push into PR branches!

File: testing/test_recwarn.py (hunk 1 above)
Comment (lesteve) [inline testing/test_recwarn.py]: Done

Comment (nicoddemus) [inline testing/test_recwarn.py]: Thanks!

File: _pytest/recwarn.py [hunk 2]
```diff
@@ -216,7 +216,8 @@ def __exit__(self, *exc_info):
         # only check if we're not currently handling an exception
         if all(a is None for a in exc_info):
             if self.expected_warning is not None:
-                if not any(r.category in self.expected_warning for r in self):
+                if not any(issubclass(r.category, exp_warning) for
```
Comment (RonnyPfannschmidt) [inline _pytest/recwarn.py]: lets use issubclass with the propper classinfo support here, no need for own loops, please check the docs

Comment (lesteve) [inline _pytest/recwarn.py]: Good point!

**Comment by nicoddemus** [comment]
Thanks a lot @lesteve!

**Comment by lesteve** [comment]
Great, thanks a lot!

**Pull request merged**